CMake: update FindCUDAToolkit.cmake, use torch::nvtx3 if present, mod…

whitesscott opened the pull request:
CMake: update FindCUDAToolkit.cmake, use torch::nvtx3 if present, modify torch/__init__.py to include libnvToolsExt.so if cuda < 12.9

Fixes #152756

Cuda-12.9 removed libnvToolsExt.so.* and is now pure header nvtx3.

Modify torch/__init__.py to include libnvToolsExt.so only if cuda < 12.9

pytorch/cmake/Modules/FindCUDAToolkit.cmake is old enough to not have any reference to CUDA::**nvtx3**.  I updated to the current version FindCUDAToolkit.cmake.

whitesscott commented:
@pytorchbot label "topic: not user facing"

Skylion007 commented on cmake/Dependencies.cmake:
```diff
@@ -1018,22 +1018,24 @@ endif()
 # ---[ nvtx
 if(USE_SYSTEM_NVTX)
-  find_path(nvtx3_dir NAMES nvtx3 PATHS ${CUDA_INCLUDE_DIRS})
+  find_path(nvtx3_dir NAMES nvtx3 PATHS /usr/local/cuda/include ${CUDA_INCLUDE_DIRS})
```
Why is this hard coded? is there not some CMake variable for this?

whitesscott replied:
mistaken testing artifact. Thank you for finding it.

johnnynunez commented:
how is it going? breaks open3d compilation @Skylion007 @eqy @ngimel

whitesscott commented:
johnnynunez , if you just want to compile open3d following should work if that's where it is failing. This is my first pr and I got too ambitious and should have just improved this rather than modifying the other related cmake files; which caused all the test failures.

pip show torch.  Edit wherever torch is installed
Not knowing cmake  hacked my ./python3.12/site-packages/torch/share/cmake/Caffe2/public/cuda.cmake

search for nvt 
substitute what is under nvToolsExt cmake if block with following and it should compile 
# nvToolsExt
if(USE_SYSTEM_NVTX)
  #find_path(nvtx3_dir NAMES nvtx3 PATHS ${CUDA_INCLUDE_DIRS}) ssw
  find_path(nvtx3_dir NAMES nvtx3 PATHS /usr/local/cuda/include ${CUDA_INCLUDE_DIRS})
#else()
#  find_path(nvtx3_dir NAMES nvtx3 PATHS "${PROJECT_SOURCE_DIR}/third_party/NVTX/c/include" NO_DEFAULT_PATH)
endif()
find_package_handle_standard_args(nvtx3 DEFAULT_MSG nvtx3_dir)
+if(NOT TARGET CUDA::nvtx3)
+  add_library(CUDA::nvtx3 INTERFACE IMPORTED)
+  target_include_directories(CUDA::nvtx3 SYSTEM INTERFACE "${CUDAToolkit_INCLUDE_DIRS}")
+  target_link_libraries(CUDA::nvtx3 INTERFACE ${CMAKE_DL_LIBS})
+endif()
#if(nvtx3_FOUND)
#  add_library(CUDA::nvtx3 INTERFACE IMPORTED)
#  target_include_directories(torch::nvtx3 INTERFACE "${nvtx3_dir}")
#  target_compile_definitions(torch::nvtx3 INTERFACE TORCH_CUDA_USE_NVTX3)
#else()
#  message(WARNING "Cannot find NVTX3, find old NVTX instead")
#  add_library(torch::nvtoolsext INTERFACE IMPORTED)
#  set_property(TARGET torch::nvtoolsext PROPERTY INTERFACE_LINK_LIBRARIES CUDA::nvToolsExt)
#endif()

johnnynunez commented:
> johnnynunez , if you just want to compile open3d following should work if that's where it is failing. This is my first pr and I got too ambitious and should have just improved this rather than modifying the other related cmake files; which caused all the test failures.
> 
> pip show torch. Edit wherever torch is installed Not knowing cmake hacked my ./python3.12/site-packages/torch/share/cmake/Caffe2/public/cuda.cmake
> 
> search for nvt substitute what is under nvToolsExt cmake if block with following and it should compile
> 
> ```
> # nvToolsExt
> if(USE_SYSTEM_NVTX)
>   #find_path(nvtx3_dir NAMES nvtx3 PATHS ${CUDA_INCLUDE_DIRS}) ssw
>   find_path(nvtx3_dir NAMES nvtx3 PATHS /usr/local/cuda/include ${CUDA_INCLUDE_DIRS})
> #else()
> #  find_path(nvtx3_dir NAMES nvtx3 PATHS "${PROJECT_SOURCE_DIR}/third_party/NVTX/c/include" NO_DEFAULT_PATH)
> endif()
> find_package_handle_standard_args(nvtx3 DEFAULT_MSG nvtx3_dir)
> +if(NOT TARGET CUDA::nvtx3)
> +  add_library(CUDA::nvtx3 INTERFACE IMPORTED)
> +  target_include_directories(CUDA::nvtx3 SYSTEM INTERFACE "${CUDAToolkit_INCLUDE_DIRS}")
> +  target_link_libraries(CUDA::nvtx3 INTERFACE ${CMAKE_DL_LIBS})
> +endif()
> #if(nvtx3_FOUND)
> #  add_library(CUDA::nvtx3 INTERFACE IMPORTED)
> #  target_include_directories(torch::nvtx3 INTERFACE "${nvtx3_dir}")
> #  target_compile_definitions(torch::nvtx3 INTERFACE TORCH_CUDA_USE_NVTX3)
> #else()
> #  message(WARNING "Cannot find NVTX3, find old NVTX instead")
> #  add_library(torch::nvtoolsext INTERFACE IMPORTED)
> #  set_property(TARGET torch::nvtoolsext PROPERTY INTERFACE_LINK_LIBRARIES CUDA::nvToolsExt)
> #endif()
> ```

Yes. I can compile modifying the files. But i would like that this PR get merged

whitesscott commented:
I believe #151583 resolves this pr. 


The remaining question would be whether  to retain or delete nvtx line, but there is a windows dll block that is a safer method to test existence rather than deletion; but that is not in this pr. 
```
torch/__init__.py   
"nvtx": "libnvToolsExt.so.*[0-9]",
```

```
        if not builtins.any(
            os.path.exists(os.path.join(p, "nvToolsExt64_1.dll")) for p in dll_paths
        ):
            nvtoolsext_dll_path = os.path.join(
                os.getenv(
                    "NVTOOLSEXT_PATH",
                    os.path.join(pfiles_path, "NVIDIA Corporation", "NvToolsExt"),
                ),
                "bin",
                "x64",
            )
        else:
            nvtoolsext_dll_path = ""
```

whitesscott commented:
I believe #151583 resolves this pr.  The following should perhaps be addressed in a separate pr?


The remaining question is which method would be better.  This pr method (top block) or a copy / modification of the existing windows `__init__.py`  logic for linux or the windows dll (bottom block) ?

`torch/__init__.py  `

```
# libnvToolsExt.so.*[0-9] is only available on CUDA < 12.9
if cuda_version:
    major, minor, *_ = map(int, cuda_version.split("."))  # ignore patch if present
    if (major, minor) < (12, 9):
        cuda_libs["nvtx"] = "libnvToolsExt.so.*[0-9]"
```



```
        if not builtins.any(
            os.path.exists(os.path.join(p, "nvToolsExt64_1.dll")) for p in dll_paths
        ):
            nvtomodifyingolsext_dll_path = os.path.join(
                os.getenv(
                    "NVTOOLSEXT_PATH",
                    os.path.join(pfiles_path, "NVIDIA Corporation", "NvToolsExt"),
                ),
                "bin",
                "x64",
            )
        else:
            nvtoolsext_dll_path = ""
```

whitesscott committed:
torch/__init__: load nvToolsExt only on CUDA < 12.9
__builtins__.int(x)  Because lintrunner didn't like the int defined somewhere in __init__.py

Yacklin commented:
@whitesscott so why you deleted it? We need it desperately! Call the reviewers!
